fluent.runtime/fluent/runtime/fallback.py

```diff
@@ -0,0 +1,83 @@
+# -*- coding: utf-8 -*-
+
+import codecs
+import os
+
+
+class FluentLocalization(object):
+    def __init__(
+        self, locales, resource_ids, resource_loader,
+        use_isolating=False,
+        bundle_class=None, functions=None,
+    ):
+        self.locales = locales
+        self.resource_ids = resource_ids
+        self.resource_loader = resource_loader
+        self.use_isolating = use_isolating
+        if bundle_class is None:
+            from fluent.runtime import FluentBundle
+            self.bundle_class = FluentBundle
+        else:
+            self.bundle_class = bundle_class
+        self.functions = functions
+        self._bundle_cache = []
+        self._bundle_it = self._iterate_bundles()
+
+    def format_value(self, id, args=None):
+        for bundle in self._bundles():
+            if not bundle.has_message(id):
+                continue
+            msg = bundle.get_message(id)
+            if not msg.value:
+                continue
+            val, errors = bundle.format_pattern(msg.value, args)
+            return val
+        return id
+
+    def _bundles(self):
+        bundle_pointer = 0
+        while True:
+            if bundle_pointer == len(self._bundle_cache):
+                try:
+                    self._bundle_cache.append(next(self._bundle_it))
+                except StopIteration:
+                    return
+            yield self._bundle_cache[bundle_pointer]
+            bundle_pointer += 1
+
+    def _iterate_bundles(self):
+        for first_loc in range(0, len(self.locales)):
+            locs = self.locales[first_loc:]
+            for resources in self.resource_loader.resources(locs[0], self.resource_ids):
+                bundle = self.bundle_class(
+                    locs, functions=self.functions, use_isolating=self.use_isolating
+                )
+                for resource in resources:
+                    bundle.add_resource(resource)
+                yield bundle
+
+
+class ResourceLoader(object):
+    def resources(self, locale, resource_ids):
+        raise NotImplementedError
+
+
+class RootedFileResourceLoader(ResourceLoader):
+    def __init__(self, root):
+        self.root = root
+        from fluent.runtime import FluentResource
+        self.Resource = FluentResource
+
+    def resources(self, locale, resource_ids):
+        resources = []
+        for resource_id in resource_ids:
+            path = self.localize_path(os.path.join(self.root, resource_id), locale)
+            if not os.path.isfile(path):
+                continue
+            content = codecs.open(path, 'r', 'utf-8').read()
+            resources.append(self.Resource(content))
+        if resources:
+            yield resources
```

Comment:
Why not just return the resources list here?

Reply:
or instead of making a list at all you can yield each resource as it's created.

Reply:
The trick here is that `resources()` can return several different combinations of resource lists, in some setups. I started implementing that, and then optimized to just make that possible. The easiest one (and one we might actually need, depending on deployment), is the `MultiRootResourceLoader`, which might look something like this:

```python
class MultiRootResourceLoader(ResourceLoader):
  def __init__(self, roots):
    self.loaders = [RootedFileResourceLoader(root) for root in roots]
  def resources(self, *args):
    for loader in self.loaders:
      yield from loader.resources(*args)
```

This would allow us to load from both the bedrock and the l10n repo, for example.

In Firefox, we actually allow to mix the different file sources, which makes that code even more complex.

I opted to add that complexity as needed, which is why this code looks complex, but doesn't show why ;-)

Reply:
Sure, but the same could be accomplished with something like this:

```python
import itertools

class MultiRootResourceLoader(ResourceLoader):
    def __init__(self, roots):
        self.loaders = [RootedFileResourceLoader(root) for root in roots]

    def resources(self, *args):
        return itertools.chain(*[l.resources(*args) for l in self.loaders])
```

That would still return an iterator that just gave resources one at a time.